File: ruby/src/Structural/Decorator/ChaiDecorator.rb
#- the decorator

from Tea import Tea

class ChaiDecorator(Tea):

	chaiIngredients = []

	def initialize(teaToMakeChai) :
		self.teaToMakeChai = teaToMakeChai
		self.chaiIngredients = [ 
			'bay leaf', 
			'cinnamon stick', 
			'ginger', 
			'honey', 
			'soy milk', 
			'vanilla bean' ]

	def steepTea(self) :
		return self.steepChai()

	def steepChai(self) :
		self.teaToMakeChai.steepTea()
		return self.steepChaiIngredients() + ', tea is steeping with chai'

	def steepChaiIngredients(self) :
		return ', '.join( [ (i + ' is steeping ') for i in self.chaiIngredients] )

File: csharp/src/Structural/Decorator/ChaiDecorator.cs
//ChaiDecorator.java - the decorator

namespace Structural.Decorator {

using System.Collections.Generic;

public class ChaiDecorator : Tea {
    private Tea teaToMakeChai;
    private List<string> chaiIngredients = new List<string>();
    
    public ChaiDecorator(Tea teaToMakeChai) {
        this.addTea(teaToMakeChai);
        chaiIngredients.Add("bay leaf");
        chaiIngredients.Add("cinnamon stick");
        chaiIngredients.Add("ginger");
        chaiIngredients.Add("honey");
        chaiIngredients.Add("soy milk");
        chaiIngredients.Add("vanilla bean");
    }

    private void addTea(Tea teaToMakeChaiIn) {
        this.teaToMakeChai = teaToMakeChaiIn;
    }
    
    override public string steepTea() {
        return this.steepChai();
    }

    public string steepChai() {
        teaToMakeChai.steepTea();
        return this.steepChaiIngredients();
    }    
    
    public string steepChaiIngredients() {
			List<string> retval = new List<string>();
			foreach(string ingred in chaiIngredients)
			{
				retval.Add( ingred + " is steeping");
			  }
			return string.Join(", ",retval.ToArray());
    }      
}

}
